src/simulate-swap.ts

```diff
@@ -0,0 +1,927 @@
+  } else {
+    return getNextSqrtPriceYDown(startingSqrtPrice, liquidity, amount, true)
+  }
+}
+
+const getNextSqrtPriceFromOutput = (
+  startingSqrtPrice: bigint,
+  liquidity: bigint,
+  amount: bigint,
+  xToY: boolean
+): bigint => {
+  if (xToY) {
+    return getNextSqrtPriceYDown(startingSqrtPrice, liquidity, amount, false)
+  } else {
+    return getNextSqrtPriceXUp(startingSqrtPrice, liquidity, amount, false)
+  }
+}
+
+const getNextSqrtPriceXUp = (
+  startingSqrtPrice: bigint,
+  liquidity: bigint,
+  x: bigint,
+  addX: Boolean
+): bigint => {
+  if (x === 0n) {
+    return startingSqrtPrice
+  }
+
+  const deltaSqrtPrice = rescale(liquidity, LIQUIDITY_SCALE, SQRT_PRICE_SCALE)
+
+  let denominator = 0n
+  if (addX) {
+    denominator = deltaSqrtPrice + mulDiv(startingSqrtPrice, x, TOKEN_AMOUNT_DENOMINATOR)
+  } else {
+    denominator = deltaSqrtPrice - mulDiv(startingSqrtPrice, x, TOKEN_AMOUNT_DENOMINATOR)
+  }
+
+  const nominator = mulDivUp(startingSqrtPrice, liquidity, LIQUIDITY_DENOMINATOR)
+
+  return divUp(nominator, denominator, SQRT_PRICE_DENOMINATOR)
+}
+
+const getNextSqrtPriceYDown = (
+  startingSqrtPrice: bigint,
+  liquidity: bigint,
+  y: bigint,
+  addY: Boolean
+): bigint => {
+  const numerator = rescale(y, TOKEN_AMOUNT_SCALE, SQRT_PRICE_SCALE)
+  const denominator = rescale(liquidity, LIQUIDITY_SCALE, SQRT_PRICE_SCALE)
+
+  if (addY) {
+    return startingSqrtPrice + div(numerator, denominator, SQRT_PRICE_DENOMINATOR)
+  } else {
+    return startingSqrtPrice - divUp(numerator, denominator, SQRT_PRICE_DENOMINATOR)
+  }
+}
+
+// in the OG code, the remainingAmount is declared as mut, but we modify it outside right after the call to this function
+const poolUpdateTick = (
+  pool: Pool,
+  tick: TickVariant | undefined,
+  nextSqrtPrice: bigint,
+  swapLimit: bigint,
+  remainingAmount: bigint,
+  byAmountIn: boolean,
+  xToY: boolean,
+  hasLimitingTick: boolean,
+  isLimitingTickInitialized: boolean,
+  limitingTickIndex: bigint
+): [bigint, bigint, boolean, boolean] => {
+  let hasCrossed = false
+  let stateInconsistency = false
+  let totalAmount = 0n
+
+  // if there's no tick we do not have to check for initialization
+  if (!tick || swapLimit !== nextSqrtPrice) {
+    pool.currentTickIndex = getTickAtSqrtPrice(nextSqrtPrice, pool.poolKey.feeTier.tickSpacing)
+    return [totalAmount, remainingAmount, hasCrossed, stateInconsistency]
+  }
+
+  const isEnoughAmountToCross = isEnoughAmountToChangePrice(
+    remainingAmount,
+    nextSqrtPrice,
+    pool.liquidity,
+    pool.poolKey.feeTier.fee.v,
+    byAmountIn,
+    xToY
+  )
+
+  if (tick && hasLimitingTick && isLimitingTickInitialized) {
+    if (!xToY || isEnoughAmountToCross) {
+      const [add, liquidityDelta] = cross(tick, pool.currentTickIndex)
+      poolCrossLiquidityUpdate(pool, add, liquidityDelta)
+      hasCrossed = true
+      if (pool.liquidity < 0n || pool.liquidity > MAX_U256) {
+        stateInconsistency = true
```

Comment:
Is it the only place where we should check if state is outdated?

Reply:
For this function, yes, see: A0's sd/src/wasm/storage/pool.rs -> line 154, returning `TrackableResult::Error` if underflow or overflow happens.